### PR TITLE
Add support for an empty `tdb_matrix`

### PR DESCRIPTION
**Pull request by jparismorgan**
### What
When we add type-erased APIs we will have a flow where we need to create an empty index:
```
def create(
    uri: str,
    dimensions: int,
    vector_type: np.dtype,
    id_type: np.dtype = np.uint32,
    adjacency_row_index_type: np.dtype = np.uint32,
    group_exists: bool = False,
    config: Optional[Mapping[str, Any]] = None,
    storage_version: str = STORAGE_VERSION,
    **kwargs,
) -> VamanaIndex:
      ctx = vspy.Ctx(config)
      index = vspy.IndexVamana(
          feature_type=np.dtype(vector_type).name, 
          id_type=np.dtype(id_type).name, 
          adjacency_row_index_type=np.dtype(adjacency_row_index_type).name, 
          dimension=dimensions,
        )
      index.write_index(ctx, uri)
      return VamanaIndex(uri=uri, config=config, memory_budget=1000000) # this is a Python VamanaIndex object.
```
When we do that we need to be able to write empty arrays and vectors. This writing of empty arrays and vectors works, but when we try to read them back we use code like this for both vectors and matrices to fill in the returned data:
```
  if (start_pos == 0) {
    start_pos = array_rows_.template domain<domain_type>().first;
  }
  if (end_pos == 0) {
    end_pos = array_rows_.template domain<domain_type>().second + 1;
  }
```
Because of this we then try to run the query:
```
tiledb::Query query(ctx, *array_);
  query.set_subarray(subarray).set_data_buffer(
      attr_name, data_.data(), vec_rows_);
  tiledb_helpers::submit_query(tdb_func__, uri, query);
```
But we crash because `data_.data()` is actually empty and so there is no where to write the data to.

The fix we make here is to disambiguate between a matrix that we should read all the rows / cols from and a matrix to read nothing from - i.e. we not support an empty `tdbMatrix`.

Note that the original version of this PR was #254, but we move over to here to split up code for supporting an empty matrix vs an empty vector.

### Testing
* Added new unit tests.
* Existing unit tests pass.

### TODO
In future PRs we will:
* Update `tdb_matrix` to use the correct types for first_row/last_row/first_col/last_col: https://app.shortcut.com/tiledb-inc/story/42696

### CODE REVIEW

File: src/include/detail/linalg/tdb_matrix.h [hunk 1]
```diff
@@ -228,15 +253,17 @@ class tdbBlockedMatrix : public MatrixBase {
 
     /* The size of the array may not be the size of domain.  Use non-zero value
      * if set in constructor */
-    if (last_row_ == 0) {
-      last_row_ =
-          (row_domain.template domain<row_domain_type>().second -
-           row_domain.template domain<row_domain_type>().first + 1);
+    if (!last_row.has_value()) {
```
Comment (lums658) [inline src/include/detail/linalg/tdb_matrix.h:256]: This is where we used to test for zero -- now that we are passing an `std::optional`, for `first_row` and `first_col` we should also pass an optional to the constructor where we used to pass the zero sentinel (per the above comments).

File: src/include/test/unit_tdb_matrix.cc [hunk 2]
```diff
@@ -271,3 +271,49 @@ TEST_CASE("tdb_matrix: MatrixBase template parameter", "[tdb_matrix]") {
     }
   }
 }
+
+TEST_CASE("tdb_matrix: empty matrix", "[tdb_matrix]") {
```
Comment (lums658) [inline src/include/test/unit_tdb_matrix.cc:275]: We should add a test for oddball `start` and `stop` as well as for "half full" - `(start, std::nullopt`) or (`std::null opt, stop`) where the array has non-zero `start`.

File: apis/python/test/test_module.py [hunk 3]
```diff
@@ -13,12 +13,32 @@ def test_tdbMatrix(tmpdir):
...
 
+    # Read all rows and cols by specifying how many rows and cols there are.
+    m = vspy.tdbColMajorMatrix_f32(ctx, p, 0, data.shape[0], 0, data.shape[1], 0, 0)
+    m.load()
+    m_array = np.array(m)
+    assert m_array.shape == data.shape
+    assert np.array_equal(m_array, data)
+
+    # Read all rows and no cols.
+    m_no_fill = vspy.tdbColMajorMatrix_f32(ctx, p, 0, None, 0, 0, 0, 0)
```
Comment (lums658) [inline apis/python/test/test_module.py:31]: With the `std::optional` approach there isn't really a way to specify a length, i.e., a `std::nullopt` plus an offset, rather we can only specify an actual `stop`.  If we don't know the start of the domain, we can't compute the size.  This is also a problem for vectors and is also a problem in previous approach with `0` is a sentinel value.  This doesn't seem to have been a problem since all the domains have been starting at `0` (so far) and so `stop` and `size` happen to coincide.

But we similarly don't know the size if we go in the other half-full direction (even if we start at `0`) so this might not be an issue.  We should definitely *not* define the arguments to mean origin + size.

**Review by lums658 — CHANGES_REQUESTED** [review]
Main comments.
* We need a consistent definition (and documentation) of what `nullopt` means and what `tdbMatrix` indices mean relative to tileDB array indices.  How these are defined for vectors and `tdbMatrix` are not consistent (this was true even with sentinel values of `0`).
  * What I propose is in an extensive comment below, but basically the idea is that the indices passed in the constructor are always relative to the domain of the TileDB array.
  * I think we are still using `0` where we mean `nullopt`.
* There is also a question about what we mean by empty array vs empty `Matrix`.  When we `create_empty_for_matrix`, the TileDB array that is created isn't really empty, in the sense that the domain is not empty.  Rather, the `Matrix` takes an empty view into that array.  (In fact, our index classes create "empty" arrays with domains that go from `0` to `MAXINT`.)
There is lots of text in my comments, but I think the actual changes in the code are quite small.  It is mostly about getting a solid mental model for indexing and making `read_vector` consistent with `tdbMatrix` constructors.

File: apis/python/src/tiledb/vector_search/module.cc [hunk 4]
```diff
@@ -395,14 +395,14 @@ static void declareColMajorMatrixSubclass(
...
-          size_t,
-          size_t,
-          size_t,
-          size_t,
-          size_t,
-          uint64_t>(),
+          const Ctx&,             // ctx
+          std::string,            // uri
+          size_t,                 // first_row
+          std::optional<size_t>,  // last_row
```
[the remark this answers is not in the document]
Comment (jparismorgan) [reply apis/python/src/tiledb/vector_search/module.cc:401]: Yes I agree, this was explained in the `TODO` section of this PR, where it also links to a story.

**Review by lums658 — APPROVED** [review]
Per discussion between @lums658 and @jparismorgan, the various subtle issues in my comments regarding start/stop etc are deferred to a subsequent PR.

Conditionally approved based on discussion, with a few small, non-subtle, fixes to be applied prior to merge.

**Comment by jparismorgan** [comment]
> Per discussion between @lums658 and @jparismorgan, the various subtle issues in my comments regarding start/stop etc are deferred to a subsequent PR.

Thanks, created a story for this here: https://app.shortcut.com/tiledb-inc/story/42770

File: src/include/test/unit_tdb_matrix.cc [hunk 5]
```diff
@@ -271,3 +271,49 @@ TEST_CASE("tdb_matrix: MatrixBase template parameter", "[tdb_matrix]") {
...
+  SECTION("filled") {
+    auto X = tdbColMajorMatrix<float>(ctx, tmp_matrix_uri);
+    X.load();
+    CHECK(X.num_cols() == matrix_domain);
+    CHECK(num_vectors(X) == matrix_domain);
+    CHECK(X.num_rows() == matrix_dimension);
+    CHECK(dimension(X) == matrix_dimension);
+
+    auto Y = tdbColMajorMatrix<float>(std::move(X));
+    CHECK(Y.num_cols() == X.num_cols());
```
[the remark this answers is not in the document]
Comment (jparismorgan) [reply src/include/test/unit_tdb_matrix.cc:335]: Sure, done, though it shouldn't affect things here because all these checks for for variables initialized in the constructor. That said, it is a good to verify that `Y.load();` doesn't change anything, so I updated to that.

File: src/include/test/unit_tdb_matrix_with_ids.cc [hunk 6]
```diff
@@ -241,3 +241,63 @@ TEST_CASE("tdb_matrix_with_ids: load from uri", "[tdb_matrix_with_ids]") {
...
+    X.load();
+    CHECK(X.num_cols() == matrix_domain);
+    CHECK(num_vectors(X) == matrix_domain);
+    CHECK(X.num_rows() == matrix_dimension);
+    CHECK(dimension(X) == matrix_dimension);
+    CHECK(X.num_ids() == matrix_domain);
+    CHECK(X.ids().size() == matrix_domain);
+
+    auto Y = tdbColMajorMatrixWithIds<float>(std::move(X));
+    CHECK(Y.num_cols() == X.num_cols());
```
[the remark this answers is not in the document]
Comment (jparismorgan) [reply src/include/test/unit_tdb_matrix_with_ids.cc:296]: Sure, done.
